### ISSUE TITLE
Whitelist, greylist or blacklist sender address, IP or domain

### ISSUE THREAD

**Issue by zxpower** [Mailu/Mailu#440]
It would be as a good addition to suite if it would be possible to whitelist, greylist or blacklist on global, per domain or per user level an item, e.g. a domain, an address, an IP.

**Comment by kaiyou** [comment]
Plus, as mentioned in the chat, a way to prioritize the action. I was thinking gray, then black, then white. And an additional attribute to "enforce" an action despite overloading, with a priority order as follows: enforced global, enforced domain, user, domain, global.

From a ui point of view, it would require views to display rules from higher levels.

**Comment by kaiyou** [comment]
Another addition: one should be able to specify external lists, like DNSBL.

**Comment by kaiyou** [comment]
This overrides #336 and #349

**Comment by kaiyou** [comment]
My hesitation at the moment, please pitch in if you have an opinion on the matter.

In the coming weeks, implementing dynamic filters in Postfix will be feasible. But this excludes more complex behaviors like dnsbl.

Should we start implementing things in Postfix and later switch to rspamd for more complex behaviors or should we wait for rspamd to be able to query the admin container?

**Comment by bpasson** [comment]
Any news on this? Currently calendar invites from gmail.com is on is blocked by bl.rspamd.com blacklist. For good reasons I assume, but it would be nice to have a way to whitelist trusted senders (email addresses) to be able to accept invites. Is it possible to add something to the rspamd module like suggested in https://github.com/rspamd/rspamd/issues/881?

'''
WHITELIST_SENDER_DOMAIN {
    type = "from";
    filter = "email:domain";
    map = "/etc/rspamd/whitelist.sender.domain.map";
    score = -6.0
  }
'''

Is there a work around for now? Something using the /etc/rspamd/override.d folder?

**Comment by Perflyst** [comment]
Is there currently a way to do **not** filter to specific domains? For example everything which is sent to domain.tld should **not** be filtered and should just go straight to the INBOX

**Comment by kaiyou** [comment]
There is not, but I agree there should definitely be one. Will keep working on this.

**Comment by dennorske** [comment]
I am not sure if this is directly involved with the contents of this issue, but it seems related: 
Setting up a relay server seems to voilate SPF configuration for the sender domain, such as google, and the antispam service immediately flags this with 3.5 negative-loaded points to the email.
Would there already be any form for whitelisting in this regard (domain/IP wise) or is this a part of the work that has to be done? 
Thank you

**Comment by kaiyou** [comment]
You mean an incoming relay server? In that case unfortunately the only way to properly handle spf would be to do it on the first hop, or to parse server headers if you trust your network enough so emails are only deliverered to Mailu from the relay.

**Comment by micw** [comment]
If one would whitelist a domain (or even a single email adddress), wouldn't this make any other spam checks obsolete? Since domain or addresses can easily be forged, spam can be send using this domain or address from every source.

**Comment by zxpower** [comment]
Is there any updates with this ticket?

Because now I'm facing the case when I need to [whitelist IP ranges for MailChimp](https://mailchimp.com/help/my-subscribers-arent-receiving-my-campaigns/) - and I still don't know how to do it... 🙄

**Comment by zxpower** [comment]
Just for the info - here is a small helper [gist](https://gist.github.com/ThomasLeister/f41adad98bb46d0c8418de50b5efb4a0) by Thomas Leister to get your `rspamd` whitelist IPs.

**Comment by elandorr** [comment]
I would also like to request adding at least basic white/blacklists for mails/IPs/domains. It would be great to have such a feature both globally and per-user.

Thank you for your efforts.

**Comment by Nebukadneza** [comment]
FYI: https://github.com/Mailu/Mailu/pull/1586/files#diff-ea4d6279fd57bb5921b88d4904f4352bR71  adds documentation towards blacklisting. With this in mind, I’d also like to copypaste our current appeal for feedback.

Hi There,

The `Mailu`-Project is currently in a bit of a bind! We are short on man-power, and we need to judge if it is possible for us to put in some work on this issue.

To help with that, we are currently trying to find out which issues are actively keeping users from using `Mailu`, which issues have someone who want to work on them — and which issues may be less important. These a less important ones could be discarded for the time being, until the project is in a more stable and regular state once again.

In order for us to better assess this, it would be helpful if you could put a **reaction on this post** (use the :smiley: icon to the top-right).
- 👍️ if you **need this** to be able to use Mailu. Ideally, you’d also be able to test this on your installation, and provide feedback …
- 🎉 if you find it a **nice bonus**, but no deal-breaker
- 🚀 if you want to **work on it yourself**!
We want to keep this **voting open for 2 weeks** from now, so please help out!

**Comment by unixfox** [comment]
Unstale

**Comment by Diman0** [comment]
It is unclear for me what questions remain unanswered?
It is already possible to configure blacklisting, whitelisting or greylisting in Rspamd. 
Rspamd is the antispam solution used in Mailu.

See [this link (click me)](https://mailu.io/master/antispam.html#how-can-i-block-emails-from-a-domain) for more information. 
Does anyone still have more questions after checking the example from the link?

**Comment by bmiller59** [comment]
@Diman0 That link was helpful! This example does not talk about whitelisting, but I was able to figure that out from 
* The mailu documentation you referenced: https://mailu.io/master/antispam.html#how-can-i-block-emails-from-a-domain
* This gist: https://gist.github.com/ThomasLeister/f41adad98bb46d0c8418de50b5efb4a0

I created the `multimap.conf` file in the overrides folder as described in the mailu doc above.
```
  WHITELIST_SENDER_DOMAIN {
      type = "from";
      filter = "email:domain";
      map = "/etc/rspamd/override.d/whitelist.sender.domain.map";
      score = -6.0
  }

```

And then set the domain sender whitelist to use file: `/etc/rspamd/override.d/whitelist.sender.domain.map` in the mailu overrides folder, as shown above.

I was pleased to see that this map file was automatically included in the "Lists" area of the Configuration tab of the rspamd admin interface (accessible via the mailu admin interface via the "Antispam" link), so that I could manage the list there easily without having to log on to my email server!

(I added these details in case they help anyone else who finds this thread.)

**Comment by witten** [comment]
It looks like there's now a greylist whitelist directly in Rspamd...? After logging into Mailu's web UI and going to Rspamd's configuration, under "Lists" there's a `/etc/rspamd/local.d/maps.d/greylist-whitelist-domains.inc` which I was able to edit and add domains to directly in the UI.

EDIT: However, changes made to that file don't appear to survive restarts... Maybe because it's not persisted anywhere outside of the container.
